### PR TITLE
Handle read returning 0

### PR DESCRIPTION
**Pull request by keithalucas**
read returning 0 for a SOCK_STREAM indicates the other side of a socket has disconnected.    When this happens, read will repeatedly return 0 and cause this function to use 100% CPU.  Returning something not equal to len in this function will notify the caller that there is an issue.

https://github.com/longhorn/longhorn/issues/3636

Signed-off-by: Keith Lucas <keith.lucas@suse.com>

### CODE REVIEW

**Comment by innobead** [comment]
cc @derekbit @shuo-wu

**Review by shuo-wu — APPROVED** [review]
LGTM.
Do we need to do anything more for the socket close detection? Can the error triggered by this modified read_full make the caller aware of the socket issue?

File: src/longhorn_rpc_protocol.c [hunk 1]
```diff
@@ -11,7 +11,7 @@ int read_full(int fd, void *buf, int len) {
 
         while (readed < len) {
                 ret = read(fd, buf + readed, len - readed);
-                if (ret < 0) {
+                if (ret <= 0) {
                         return ret;
```
Comment (derekbit) [inline src/longhorn_rpc_protocol.c:15]: Not sure if `return 0` is safe enough.
Is the case possible?
```
len = 8192
1st read = 4096
2st read = 0
3st read = 4096 
```

Comment (keithalucas) [reply src/longhorn_rpc_protocol.c]: No, if the socket connection is open, `read` won't return 0 if the next `read` could read something and return a value greater than zero.

This is uses blocking I/O so `read` will wait until there is something to read.    So the return value of zero effectively means end of file (EOF).   If you were using `read` on a file descriptor that corresponded to a file on disk, the return value of zero will mean you reached the end of that file and every subsequent `read` would also return zero.

The only scenario where `read` could return without reading something on a blocking file descriptor would be if it returned -1 and set `errno` to `EINTR` to indicate the system call was interrupted.  I handle this now. 

I also return `readed` when `ret` equal 0 now and I renamed `readed` to `nread`, both on @yasker 's recommendation.

**Comment by keithalucas** [comment]
> LGTM. Do we need to do anything more for the socket close detection? Can the error triggered by this modified read_full make the caller aware of the socket issue?

The called checks if `read_full` returns the `len` requested, so when this condition occurs, the connection closes.  We print out:
```
receive_msg: fail to read magic version
response_process: Receive response returned error
lh_client_close_conn: Closing connection
```

**Review by derekbit — APPROVED** [review]
LGTM.

File: src/longhorn_rpc_protocol.c (hunk 1 above)
Comment (derekbit) [reply src/longhorn_rpc_protocol.c]: > I also return readed when ret equal 0 now and I renamed readed to nread, both on @yasker 's recommendation.

Yep, returning `readed` makes sense now.

**Review by innobead — APPROVED** [review]
LGMT. Who can help with the merge?

**Comment by joshimoo** [comment]
@keithalucas good analysis, good fix for also dealing with the syscall interruption `EINTR`
I might have considered moving the 0 case into the first case to help with pipelining, since you would only have a single branch, that will be taken 99% of the time, but gcc might end up optimizing that anyway :)

**Pull request merged**